utils/kubernetes/pod_service_mapper.py

```diff
@@ -0,0 +1,187 @@
+# (C) Datadog, Inc. 2015-2016
+# All rights reserved
+# Licensed under Simplified BSD License (see LICENSE)
+
+# stdlib
+from collections import defaultdict
+import logging
+
+log = logging.getLogger('collector')
+
+
+class PodServiceMapper:
+    def __init__(self, kubeutil_object):
+        """
+        Create a new service PodServiceMapper
+        The apiserver requests are routed through the given KubeUtil instance
+        """
+        self.kube = kubeutil_object
+        self._event_retriever = self.kube.get_event_retriever(kinds=['Service'])
+        self._service_cache_selectors = defaultdict(dict)   # {service_uid:{selectors}}
+        self._service_cache_names = {}                      # {service_uid:service_name
+        self._service_cache_invalidated = False             # True to trigger service parsing
+        self._pod_labels_cache = defaultdict(dict)          # {pod_uid:{label}}
+        self._pod_services_mapping = defaultdict(list)      # {pod_uid:[service_uid]}
+
+        # Consume past events
+        self.check_services_cache_freshness()
+        self._service_cache_invalidated = True
+
+    def _fill_services_cache(self):
+        """
+        Get the list of services from the kubelet API and store the label selector dicts.
+        The cache is to be invalidated by the user class by calling check_services_cache_freshness
+        """
+        try:
+            reply = self.kube.retrieve_json_auth(self.kube.kubernetes_api_url + '/services')
+            self._service_cache_selectors = defaultdict(dict)
+            self._service_cache_names = {}
+            for service in reply.get('items', []):
+                uid = service.get('metadata', {}).get('uid', '')
+                name = service.get('metadata', {}).get('name', '')
+                selector = service.get('spec', {}).get('selector', {})
+                if uid == '' or len(selector) == 0:
+                    continue
+                self._service_cache_names[uid] = name
+                self._service_cache_selectors[uid] = selector
+            self._service_cache_invalidated = False
+        except Exception as e:
+            log.exception('Unable to read service list from apiserver: %s', e)
+            self._service_cache_selectors = defaultdict(dict)
+            self._service_cache_names = {}
```

Comment:
That exception catch would be a very last resort (as we are only doing safe dict lookups). In the case it happens, I think it's better to have no service tagging than only half of them tagged (easier to troubleshoot).